File: examples/openai-realtime/test_smoke_shape.mjs
#!/usr/bin/env node
// Smoke test for examples/openai-realtime/main.js protocol shape.
//
// The reference relay handles three production-grade behaviours: ephemeral
// session tokens, sample-rate matching (16 → 24 kHz on the way in),
// barge-in via `response.cancel` (not mute), and function calling. This
// test verifies the wire-protocol surface without running anything.
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const src = fs.readFileSync(path.join(__dirname, 'main.js'), 'utf8');

const checks = [
  // SAA surface.
  ['imports AttentionClient', /import\s*\{\s*AttentionClient\s*\}/.test(src)],
  ['subscribes to speechReady', /saa(Client)?\.on\("speechReady"|"speechReady"/.test(src)],

  // OpenAI Realtime surface.
  ['references gpt-realtime model', /model=gpt-realtime|"gpt-realtime"/.test(src)],
  ['uses input_audio_buffer.append', /input_audio_buffer\.append/.test(src)],
  ['uses input_audio_buffer.commit', /input_audio_buffer\.commit/.test(src)],
  ['uses response.create', /response\.create/.test(src)],

  // Ephemeral-token + browser-direct dual mode.
  ['supports ephemeral client_secret', /client_secret/.test(src)],
  ['supports browser-direct subprotocol', /openai-insecure-api-key/.test(src)],

  // Sample-rate matching (16 kHz SAA ↔ 24 kHz OpenAI).
  ['declares SAA_INPUT_RATE 16 kHz', /SAA_INPUT_RATE\s*=\s*16000/.test(src)],
  ['declares OPENAI_RATE 24 kHz', /OPENAI_RATE\s*=\s*24000/.test(src)],
  ['playback at OPENAI_RATE to match model output',
    /sampleRate:\s*OPENAI_RATE|sampleRate:\s*24000/.test(src)],

  // Barge-in design (no mute(), we cancel + drop queued playback instead).
  ['markResponding(true) when agent starts', /markResponding\??\.?\(true\)/.test(src)],
  ['markResponding(false) when agent stops', /markResponding\??\.?\(false\)/.test(src)],
  ['barge-in via response.cancel', /response\.cancel/.test(src)],
  ['drops queued playback on barge-in', /stopPlayback\(\)/.test(src)],

  // Function calling.
  ['accumulates function_call_arguments.delta',
    /response\.function_call_arguments\.delta/.test(src)],
  ['handles function_call_arguments.done',
    /response\.function_call_arguments\.done/.test(src)],
  ['emits function_call_output via conversation.item.create',
    /conversation\.item\.create[\s\S]*function_call_output/.test(src)],
  ['registers get_weather tool', /"get_weather"/.test(src)],
  ['registers set_timer tool', /"set_timer"/.test(src)],

  // Transcripts + UX.
  ['shows user transcripts',
    /conversation\.item\.input_audio_transcription\.completed/.test(src)],
  ['handles response.audio.delta', /response\.audio\.delta/.test(src)],
  ['handles response.done', /response\.done/.test(src)],
];

let fail = 0;
for (const [name, ok] of checks) {
  if (ok) console.log(`✓ ${name}`);
  else { console.log(`✗ ${name}`); fail++; }
}
console.log(`openai-realtime shape: ${checks.length - fail}/${checks.length} pass`);
if (fail > 0) process.exit(1);

File: examples/openai-realtime/test_smoke_shape.py
"""Smoke test for examples/openai-realtime/server.py protocol shape.

Verifies the FastAPI relay's contract without hitting any network or
requiring fastapi/websockets/saa to be installed:

  - POST /session declared, returns ``client_secret`` from OpenAI sessions API
  - WebSocket /twilio declared with subprotocol ``audio.twilio.com``
  - GET /health declared
  - StaticFiles mount at "/" so the browser bundle is served alongside the API
  - Twilio bridge bridges µ-law 8 kHz ↔ PCM16 16 kHz ↔ PCM16 24 kHz
  - SAA client opened with enable_audio=False, enable_video=False on the
    server side (the v1.0 relay shim)
  - on_speech_ready hop into asyncio loop
  - mark_responding(True/False) drive SAA suppression during agent playback
  - Barge-in: response.cancel sent before forwarding a new utterance

No live LiveKit / OpenAI / Twilio server required.
"""
from __future__ import annotations

import ast
import pathlib
import sys

ROOT = pathlib.Path(__file__).resolve().parent
SERVER = ROOT / "server.py"


def main() -> int:
    src = SERVER.read_text()
    try:
        tree = ast.parse(src, filename=str(SERVER))
    except SyntaxError as exc:
        print(f"✗ syntax error: {exc}", file=sys.stderr)
        return 1

    failures: list[str] = []

    def need(label: str, ok: bool) -> None:
        if not ok:
            failures.append(label)

    # ── routes ───────────────────────────────────────────────────────────────────
    need('POST /session route', '@app.post("/session")' in src)
    need('GET /health route', '@app.get("/health")' in src)
    need('WebSocket /twilio route', '@app.websocket("/twilio")' in src)
    need('StaticFiles mount serves browser bundle',
         'StaticFiles(' in src and 'html=True' in src)

    # ── OpenAI bits ──────────────────────────────────────────────────────────────────
    need('mints ephemeral session against /v1/realtime/sessions',
         '/v1/realtime/sessions' in src)
    need('uses Authorization: Bearer for sessions API',
         'Bearer ' in src and 'Authorization' in src)
    need('OpenAI-Beta: realtime=v1 header',
         'OpenAI-Beta' in src and 'realtime=v1' in src)
    need('opens Realtime WS', 'wss://api.openai.com/v1/realtime' in src)
    need('Realtime WS uses realtime + insecure-api-key + beta-v1 subprotocols',
         '"realtime"' in src
         and 'openai-insecure-api-key' in src
         and 'openai-beta.realtime-v1' in src)
    need('session.update pins pcm16 input + output',
         '"input_audio_format": "pcm16"' in src
         and '"output_audio_format": "pcm16"' in src)
    need('session.update disables OpenAI VAD (turn_detection=None)',
         '"turn_detection": None' in src)
    need('forwards utterance via input_audio_buffer.append',
         '"input_audio_buffer.append"' in src)
    need('commits + creates response on each utterance',
         '"input_audio_buffer.commit"' in src and '"response.create"' in src)

    # ── Twilio bridge ───────────────────────────────────────────────────────────────
    need('accepts Twilio audio.twilio.com subprotocol',
         'subprotocol="audio.twilio.com"' in src)
    need('handles Twilio start event for streamSid', '"start"' in src and 'streamSid' in src)
    need('handles Twilio media frames', '"media"' in src)
    need('handles Twilio stop event', '"stop"' in src)
    need('responds with media events to Twilio',
         '"event": "media"' in src or "'event': 'media'" in src)

    # ── audio plumbing ────────────────────────────────────────────────────────────────
    need('upsamples µ-law 8 kHz → PCM16 16 kHz', '_ulaw8k_to_pcm16_16k_b64' in src)
    need('downsamples PCM16 24 kHz → µ-law 8 kHz', '_pcm16_24k_to_ulaw8k_b64' in src)
    need('resamples SAA 16 kHz → OpenAI 24 kHz on the way in',
         'audioop.ratecv' in src and 'SAA_RATE' in src and 'OPENAI_RATE' in src)
    need('uses ulaw2lin for Twilio decode', 'audioop.ulaw2lin' in src)
    need('uses lin2ulaw for Twilio encode', 'audioop.lin2ulaw' in src)

    # ── SAA bits ─────────────────────────────────────────────────────────────────────
    need('AttentionClient created with enable_audio=False',
         'enable_audio=False' in src)
    need('AttentionClient created with enable_video=False',
         'enable_video=False' in src)
    need('on_speech_ready handler bridges to asyncio',
         'on_speech_ready' in src and 'run_coroutine_threadsafe' in src)
    need('mark_responding(True/False) used during agent playback',
         'mark_responding(True)' in src and 'mark_responding(False)' in src)
    need('barge-in via response.cancel before forwarding new turn',
         '"response.cancel"' in src or "'response.cancel'" in src)

    if failures:
        for f in failures:
            print(f"✗ {f}", file=sys.stderr)
        return 1

    # Sanity: confirm the AST parsed and at least the documented helpers exist.
    fns = {
        n.name for n in ast.walk(tree)
        if isinstance(n, (ast.FunctionDef, ast.AsyncFunctionDef))
    }
    for required in (
        'mint_session',
        'twilio_bridge',
        '_open_openai_realtime',
        '_twilio_inbound_loop',
        '_openai_inbound_loop',
        '_forward_to_openai',
        '_ulaw8k_to_pcm16_16k_b64',
        '_pcm16_24k_to_ulaw8k_b64',
    ):
        if required not in fns:
            print(f"✗ missing function: {required}", file=sys.stderr)
            return 1

    print(
        '✓ openai-realtime server shape: routes + ephemeral mint + '
        'twilio bridge + SAA gate + barge-in'
    )
    return 0


if __name__ == '__main__':
    sys.exit(main())
